File: backend/app/models/customer.py
from sqlalchemy import Column, String, Date, Text, ForeignKey
from sqlalchemy.orm import relationship
from datetime import datetime

from app.models.base import BaseModel


class Customer(BaseModel):
    """客户表"""
    __tablename__ = "customers"
    
    # 基本信息
    code = Column(String(20), unique=True, nullable=False, comment="客户编码")
    name = Column(String(200), nullable=False, comment="客户名称")
    short_name = Column(String(50), comment="客户简称")
    
    # 企业信息
    company_type = Column(String(50), comment="企业类型")
    industry = Column(String(50), comment="所属行业")
    scale = Column(String(20), comment="企业规模")
    
    # 税务信息
    tax_no = Column(String(20), unique=True, comment="统一社会信用代码/税号")
    tax_type = Column(String(20), default="一般纳税人", comment="纳税人类型")
    
    # 联系信息
    phone = Column(String(50), comment="联系电话")
    email = Column(String(100), comment="联系邮箱")
    fax = Column(String(50), comment="传真")
    website = Column(String(200), comment="公司网站")
    
    # 状态
    status = Column(String(20), default="active", comment="状态")
    
    # 服务信息
    service_start_date = Column(Date, comment="服务开始日期")
    service_end_date = Column(Date, comment="服务结束日期")
    assigned_accountant_id = Column(ForeignKey("users.id"), comment="负责会计")
    
    # 备注
    remark = Column(Text, comment="备注")
    
    # 审计字段
    created_by = Column(ForeignKey("users.id"))
    
    # 关系
    bills = relationship("Bill", back_populates="customer")
    vouchers = relationship("Voucher", back_populates="customer")
    bank_flows = relationship("BankFlow", back_populates="customer")
    contacts = relationship("CustomerContact", back_populates="customer", cascade="all, delete-orphan")
    addresses = relationship("CustomerAddress", back_populates="customer", cascade="all, delete-orphan")
    invoice_infos = relationship("CustomerInvoiceInfo", back_populates="customer", cascade="all, delete-orphan")

File: frontend/src/api/customer.ts
import api from './index'

export interface Customer {
  id: string
  code: string
  name: string
  short_name?: string
  company_type?: string
  industry?: string
  scale?: string
  tax_no?: string
  tax_type?: string
  phone?: string
  email?: string
  fax?: string
  website?: string
  status: string
  service_start_date?: string
  service_end_date?: string
  assigned_accountant_id?: string
  remark?: string
  created_at: string
}

export interface CustomerContact {
  id?: string
  name: string
  title?: string
  department?: string
  phone?: string
  tel?: string
  email?: string
  wechat?: string
  is_primary: boolean
  remark?: string
}

export interface CustomerAddress {
  id?: string
  address_type: string
  province?: string
  city?: string
  district?: string
  detail?: string
  postcode?: string
  is_primary: boolean
}

export interface CustomerInvoiceInfo {
  id?: string
  title: string
  tax_no: string
  address?: string
  phone?: string
  bank_name?: string
  bank_account?: string
  is_default: boolean
}

export const customerApi = {
  // 获取列表
  getList: (params: {
    page?: number
    page_size?: number
    keyword?: string
    status?: string
    industry?: string
    assigned_accountant_id?: string
  }) => api.get('/customers', { params }),

  // 获取详情
  getDetail: (id: string) => api.get(`/customers/${id}`),

  // 创建客户
  create: (data: Partial<Customer>) => api.post('/customers', data),

  // 更新客户
  update: (id: string, data: Partial<Customer>) => api.put(`/customers/${id}`, data),

  // 删除客户
  delete: (id: string) => api.delete(`/customers/${id}`),

  // 添加联系人
  addContact: (customerId: string, data: CustomerContact) => 
    api.post(`/customers/${customerId}/contacts`, data),

  // 删除联系人
  deleteContact: (contactId: string) => 
    api.delete(`/customers/contacts/${contactId}`),

  // 添加地址
  addAddress: (customerId: string, data: CustomerAddress) => 
    api.post(`/customers/${customerId}/addresses`, data),

  // 删除地址
  deleteAddress: (addressId: string) => 
    api.delete(`/customers/addresses/${addressId}`),

  // 添加开票信息
  addInvoiceInfo: (customerId: string, data: CustomerInvoiceInfo) => 
    api.post(`/customers/${customerId}/invoice-infos`, data),

  // 删除开票信息
  deleteInvoiceInfo: (infoId: string) => 
    api.delete(`/customers/invoice-infos/${infoId}`)
}
